File: languages/typescript/02-core/file-io-basics/example/score-report.ts
export type ScoreRecord = {
    name: string;
    score: number;
};

export function parseScoreRow(line: string): ScoreRecord | null {
    // Parse before touching the filesystem so malformed rows can be tested directly.
    const parts = line.trim().split(/\s+/);
    if (parts.length < 2) {
        return null;
    }

    const rawScore = parts.at(-1) ?? "";
    if (!/^[+-]?\d+$/.test(rawScore)) {
        return null;
    }
    const score = Number(rawScore);
    const name = parts.slice(0, -1).join(" ");
    return Number.isSafeInteger(score) && score >= 0 && score <= 100 && name
        ? { name, score }
        : null;
}

export function buildScoreReport(
    records: ScoreRecord[],
    invalidRows: number,
): string {
    const lines = [
        "Grade Report",
        `Valid records: ${records.length}`,
        `Invalid rows skipped: ${invalidRows}`,
    ];
    if (records.length === 0) {
        return [...lines, "No valid records."].join("\n");
    }

    const total = records.reduce((sum, record) => sum + record.score, 0);
    return [
        ...lines,
        `Average: ${(total / records.length).toFixed(2)}`,
        ...records.map((record) => `- ${record.name}: ${record.score}`),
    ].join("\n");
}

File: languages/go/02-core/file-io-basics/example/score_report.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type scoreRecord struct {
	name  string
	score int
}

func parseScoreRow(line string) (scoreRecord, bool) {
	// Keep parsing independent from file access so malformed rows are easy to test.
	parts := strings.Fields(line)
	if len(parts) < 2 {
		return scoreRecord{}, false
	}
	score, err := strconv.Atoi(parts[len(parts)-1])
	name := strings.Join(parts[:len(parts)-1], " ")
	if err != nil || score < 0 || score > 100 || name == "" {
		return scoreRecord{}, false
	}
	return scoreRecord{name: name, score: score}, true
}

func buildReport(records []scoreRecord, invalidRows int) string {
	lines := []string{
		"Grade Report",
		fmt.Sprintf("Valid records: %d", len(records)),
		fmt.Sprintf("Invalid rows skipped: %d", invalidRows),
	}
	if len(records) == 0 {
		return strings.Join(append(lines, "No valid records."), "\n")
	}

	total := 0
	for _, record := range records {
		total += record.score
	}
	lines = append(lines, fmt.Sprintf("Average: %.2f", float64(total)/float64(len(records))))
	for _, record := range records {
		lines = append(lines, fmt.Sprintf("- %s: %d", record.name, record.score))
	}
	return strings.Join(lines, "\n")
}
